cmake/modules/FindSYMENGINE.cmake

```diff
@@ -0,0 +1,157 @@
+
+#
+# - Try to find SymEngine
+#
+# This module exports
+#
+#   SYMENGINE_INCLUDE_DIR
+#   SYMENGINE_LIBRARY
+#   SYMENGINE_WITH_LLVM
+#
+
+SET(SYMENGINE_DIR "" CACHE PATH "An optional hint to a SymEngine installation")
+SET_IF_EMPTY(SYMENGINE_DIR "$ENV{SYMENGINE_DIR}")
+
+DEAL_II_FIND_PATH(SYMENGINE_INCLUDE_DIR
+  NAMES symengine/symengine_config.h
+  HINTS ${SYMENGINE_DIR}
+  PATH_SUFFIXES include
+)
+
+DEAL_II_FIND_LIBRARY(SYMENGINE_LIBRARY
+  NAMES symengine
+  HINTS ${SYMENGINE_DIR}
+  PATH_SUFFIXES lib${LIB_SUFFIX} lib64 lib
+)
+
+
+#
+# SymEngine overwrites the CMake module path, so we save
+# and restore it after this library is found and configured.
+#
+SET (DEAL_II_CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH})
+
+#
+# So, up to this point it was easy. Now, the tricky part.
+# Find the SymEngine package using CMake's "Config" mode.
+# From this we'll determine what collections of libraries
+# are required by SymEngine.
+# This way we'll get all of the includes and the right
+# linker line to link to the libraries that SymEngine uses.
+# The minimum supported version of SymEngine is 0.3 due
+# to a change in the basic interface.
+#
+SET(SYMENGINE_MINIMUM_REQUIRED_VERSION "0.3")
+FIND_PACKAGE(SymEngine ${SYMENGINE_MINIMUM_REQUIRED_VERSION} CONFIG REQUIRED
+  PATHS ${SYMENGINE_DIR}
+  PATH_SUFFIXES lib/cmake/symengine
+  NO_DEFAULT_PATH
+  NO_CMAKE_ENVIRONMENT_PATH
+  NO_CMAKE_PATH
+  NO_SYSTEM_ENVIRONMENT_PATH
+  NO_CMAKE_PACKAGE_REGISTRY
+  NO_CMAKE_BUILDS_PATH
+  NO_CMAKE_SYSTEM_PATH
+  NO_CMAKE_SYSTEM_PACKAGE_REGISTRY
+)
+
+
+MESSAGE("CMAKE_MODULE_PATH_AFTER: ${CMAKE_MODULE_PATH}")
+
+#
+# We'll also parse the configuration header to determine
+# which features are activated and required
+#
+FIND_FILE(SYMENGINE_SETTINGS_H symengine_config.h
+  HINTS ${SYMENGINE_DIR}
+  PATH_SUFFIXES include/symengine
+)
+
+#
+# Version check
+#
+FILE(STRINGS "${SYMENGINE_SETTINGS_H}" SYMENGINE_VERSION_MAJOR_STRING
+  REGEX "#define.*SYMENGINE_VERSION_MAJOR")
+STRING(REGEX REPLACE "^.*SYMENGINE_VERSION_MAJOR.*([0-9]+).*" "\\1"
+  SYMENGINE_VERSION_MAJOR "${SYMENGINE_VERSION_MAJOR_STRING}"
+  )
+FILE(STRINGS "${SYMENGINE_SETTINGS_H}" SYMENGINE_VERSION_MINOR_STRING
+  REGEX "#define.*SYMENGINE_MINOR_VERSION")
+STRING(REGEX REPLACE "^.*SYMENGINE_MINOR_VERSION.*([0-9]+).*" "\\1"
+  SYMENGINE_VERSION_MINOR "${SYMENGINE_VERSION_MINOR_STRING}"
+  )
+FILE(STRINGS "${SYMENGINE_SETTINGS_H}" SYMENGINE_VERSION_PATCH_STRING
+  REGEX "#define.*SYMENGINE_PATCH_VERSION")
+STRING(REGEX REPLACE "^.*SYMENGINE_PATCH_VERSION.*([0-9]+).*" "\\1"
+  SYMENGINE_VERSION_PATCH "${SYMENGINE_VERSION_PATCH_STRING}"
+  )
+SET(SYMENGINE_VERSION
+  "${SYMENGINE_VERSION_MAJOR}.${SYMENGINE_VERSION_MINOR}.${SYMENGINE_VERSION_PATCH}"
+  )
+
+IF (SYMENGINE_VERSION LESS ${SYMENGINE_MINIMUM_REQUIRED_VERSION})
+  MESSAGE(FATAL_ERROR
+    "The version of SymEngine that has been found is ${SYMENGINE_VERSION}.
+     deal.II requires at least ${SYMENGINE_MINIMUM_REQUIRED_VERSION}")
+ENDIF()
```

Comment:
Refactor this check into the `FEATURE_SYMENGINE_FIND_EXTERNAL` macro (simply returning false if the version is insufficient, have a look at `configure_2_trilinos.cmake` how this can be done - no need for a `FATAL_ERROR` here.